### PR TITLE
(TFM) Add transfer fees module

### PR DESCRIPTION
**Pull request by aliarbak**
Notes:
- The minimum fee rate is 0.01% but must be given in uint format. So, when defining fee rates:
  - 1 = 0.01%
  - 10 = 0.1%
  - 100 = 1%
  - 1000 = 10%
  - 10000 = 100%
  - It can also be 0, which means no fee will be charged
- `canComplianceBind` function is for the plug-and-play support discussed before. It checks whether the module is a token agent (and is required for the module to function). It should be called from the `addComplianceModule`
  - This is a draft function, I can update it after we decide how to proceed with the plug-and-play support.

### CODE REVIEW

File: contracts/compliance/modular/modules/TransferFeesModule.sol [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    /**
+    *  @dev Sets the fee rate and collector of the given compliance
+    *  @param _compliance is the compliance contract address
+    *  @param _rate is the rate of the fee (0.01% = 1, 1% = 100, 100% = 10000)
+    *  @param _collector is the collector wallet address
+    *  Only the owner of the token smart contract can call this function
+    *  Collector wallet address must be verified
+    */
+    function setFee(address _compliance, uint256 _rate, address _collector) external {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/TransferFeesModule.sol:105]: all owner scoped functions should be called through the ModularCompliance function `callModuleFunction` in order to standardize the way we setup modules. some exceptions can exist, if for some reason it is not possible to achieve the desired behaviour this way, but when possible it is better to do it. Please update the function to be called by the modular compliance through `callModuleFunction`

File: contracts/compliance/modular/modules/TransferFeesModule.sol [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        address tokenAddress = IModularCompliance(_compliance).getTokenBound();
+        require(msg.sender == AgentRole(tokenAddress).owner(), "only token owner can call");
+
+        if (_rate > 10000) {
+            revert FeeRateIsOutOfRange(_compliance, _rate);
+        }
+
+        if (_collector == address(0)) {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/TransferFeesModule.sol:113]: if the address is zero address `isVerified` that is called right after will return `false` therefore we can remove this check here

File: contracts/compliance/modular/modules/TransferFeesModule.sol [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        _fees[_compliance].collector = _collector;
+        emit FeeUpdated(_compliance, _rate, _collector);
+    }
+
+    /**
+    *  @dev See {IModule-moduleTransferAction}.
+    */
+    function moduleTransferAction(address _from, address _to, uint256 _value) external override onlyComplianceCall {
+        if (_from == address(0) || _to == address(0)) {
+            return;
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/TransferFeesModule.sol:132]: i believe this part is unreachable as the token contract will never call the `transferred` function with zero address as `_from` or `_to`, we can remove this check imo

**Review by Joachim-Lebrun — CHANGES_REQUESTED** [review]
- add the `isPlugAndPlay` and `canComplianceBind` after #169 is merged to develop
- remove useless checks to save gas
- modify `setFee` function to be called by the Modular Compliance contract

**Comment by Joachim-Lebrun** [comment]
just need to still add the `isPlugAndPlay` and `canComplianceBind` functions, which are both always returning `true`

**Comment by aliarbak** [comment]
Hey @Joachim-Lebrun, I need to rebase from develop, can you allow me to force-push to the transfer-fees-module branch?

**Review by Joachim-Lebrun — APPROVED** [review]

**Pull request merged**